### ISSUE TITLE
How to create topics with message.timestamp.type config set to LogAppendTime?

### ISSUE THREAD

**Issue by junaidrazzaq323** [wurstmeister/kafka-docker#269]
By default `message.timestamp.type=CreateTime`, this means that a timestamp must be added by the producer. However if a timestamp is not added then the following exception occurs in Streams Application:

`org.apache.kafka.streams.errors.StreamsException: Extracted timestamp value is negative, which is not allowed.`

We are using a third party producer so we need to set `message.timestamp.type=LogAppendTime`.
By setting it to `LogAppendTime`, a timestamp on each message is added when it is received by the broker.

**Comment by sscaling** [comment]
@junaidrazzaq323  - did you follow the information in the Readme to override the configuration?

> if you want to customize any Kafka parameters, simply add them as environment variables in docker-compose.yml, e.g. in order to increase the message.max.bytes parameter set the environment to KAFKA_MESSAGE_MAX_BYTES: 2000000. To turn off automatic topic creation set KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'false'

i.e. set the environment variable `KAFKA_MESSAGE_TIMESTAMP_TYPE=LogAppendTime`

**Comment by junaidrazzaq323** [comment]
Thanks for the help.

**Comment by junaidrazzaq323** [comment]
I have set `KAFKA_MESSAGE_TIMESTAMP_TYPE=LogAppendTime` in docker-compose. However I am getting the following error:

`org.apache.kafka.streams.errors.StreamsException: Input record ConsumerRecord(topic = o365_storage, partition = 0, offset = 0, CreateTime = -1, serialized key size = -1, serialized value size = 88, headers = RecordHeaders(headers = [], isReadOnly = false), key = null, value = {"total":3298534883328,"deleted":4094464785,"used":6371340077,"remaining":3288191993612}) has invalid (negative) timestamp. Possibly because a pre-0.10 producer client was used to write this record to Kafka without embedding a timestamp, or because the input topic was created before upgrading the Kafka cluster to 0.10+. Use a different TimestampExtractor to process this data.`

**Docker-Compose:**

```
kafka:
    container_name: kafka
    image: wurstmeister/kafka:1.0.0
    hostname: kafka
    links:
      - zookeeper
    ports:
      - "9092:9092"
    environment:
      KAFKA_ADVERTISED_HOST_NAME: kafka
      KAFKA_CREATE_TOPICS: "o365_user_activity:1:1,o365_user_activity_by_date:1:1,o365_storage:1:1,o365_storage_by_date:1:1,o365_added_to_group_by_date:1:1,o365_added_to_secure_link_by_date:1:1,o365_sharing_set_by_date:1:1,o365_role_update_by_date:1:1,o365_upload_file_by_date:1:1"
      KAFKA_MESSAGE_TIMESTAMP_TYPE: LogAppendTime
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
```
Previously when I was not on docker environment, I used to create a topic and then run the following topic alter command:

`kafka-configs --alter --entity-name o365_user_activity --entity-type topics --add-config message.timestamp.type=LogAppendTime --zookeeper localhost:2181`

Setting `message.timestamp.type` to `LogAppendTime` solved the error.

How can I resolve this?
and Is there a way to set `message.timestamp.type=LogAppendTime` for individual topics?

Thanks.

**Comment by junaidrazzaq323** [comment]
Adding KAFKA_LOG_MESSAGE_TIMESTAMP_TYPE = LogAppendTime solved the issue.

**Comment by junaidrazzaq323** [comment]
Adding KAFKA_LOG_MESSAGE_TIMESTAMP_TYPE = LogAppendTime solved the issue.

**Comment by blockloop** [comment]
> set the environment variable KAFKA_MESSAGE_TIMESTAMP_TYPE=LogAppendTime

@sscaling this no longer seems valid because I don't see this env var anywhere in the README nor do I see it used in the source code. I'm setting both `KAFKA_MESSAGE_TIMESTAMP_TYPE` and `KAFKA_LOG_MESSAGE_TIMESTAMP_TYPE` and the timestamps are still nil on my end.

Am I doing something wrong?

```
  kafka:
    image: wurstmeister/kafka:0.10.0.0
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
    ports:
        - 9092:9092
    links:
      - zookeeper
    environment:
      KAFKA_ADVERTISED_HOST_NAME: kafka
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_AUTO_CREATE_TOPICS_ENABLE: 'true'
      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
      KAFKA_MESSAGE_TIMESTAMP_TYPE: LogAppendTime
      KAFKA_LOG_MESSAGE_TIMESTAMP_TYPE: LogAppendTime
```

**Comment by sscaling** [comment]
@blockloop - see a recent comment https://github.com/wurstmeister/kafka-docker/issues/314#issuecomment-415673986. You won't find too many specific env vars as they are just translated.

It's a valid setting for that version of Kafka and using your config, it appears to work fine for me:

```
$ docker-compose up -d
$ echo "foo" | kafkacat -b kafka:9092 -t test -P

$ kafkacat -b kafka:9092 -t test -C -f "%T: [%k]: %s\n" -o beginning
1535729448463: []: foo
% Reached end of topic test [0] at offset 1
```
I see the correct setting in the kafka logs as well

```
docker-compose logs | grep timestamp
kafka_1      | 	log.message.timestamp.type = LogAppendTime
kafka_1      | 	log.message.timestamp.difference.max.ms = 9223372036854775807
```


(p.s. I don't think `message.timestamp.type` is a valid setting, ref: https://kafka.apache.org/0100/documentation.html#brokerconfigs)
(p.p.s. You should really be using the latest 0.10.x kafka release for security patches / fixes etc - `wurstmeister/kafka:2.11-0.10.2.2` - unless you can't for client compatibility reasons etc.)
